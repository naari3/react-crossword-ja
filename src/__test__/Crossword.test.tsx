import { jest } from '@jest/globals';
import React from 'react';
import ReactDom from 'react-dom';
import { act, render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import '@testing-library/jest-dom/extend-expect';

import Crossword, { CrosswordImperative } from '../Crossword';

afterEach(cleanup);

const emptyData = {
  across: {},
  down: {},
};

const simpleData = {
  across: {
    1: {
      clue: 'apple',
      answer: 'りんご',
      row: 0,
      col: 0,
    },
  },
  down: {
    2: {
      clue: 'gorilla',
      answer: 'ごりら',
      row: 0,
      col: 2,
    },
  },
};

const defaultProps = {
  data: emptyData,
  useStorage: false,
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDom.render(<Crossword {...defaultProps} />, div);
  ReactDom.unmountComponentAtNode(div);
});

it('renders Crossword component correctly', () => {
  const { container, getByText } = render(<Crossword {...defaultProps} />);
  expect(container.firstChild).toHaveClass('crossword');
  expect(getByText('ACROSS')).toHaveTextContent('ACROSS');
  expect(getByText('DOWN')).toHaveTextContent('DOWN');
});

it('renders Crossword component correctly when using storage', () => {
  const { container, getByText } = render(
    <Crossword {...defaultProps} useStorage />
  );
  expect(container.firstChild).toHaveClass('crossword');
  expect(getByText('ACROSS')).toHaveTextContent('ACROSS');
  expect(getByText('DOWN')).toHaveTextContent('DOWN');
});

it('matches snapshot', () => {
  const tree = renderer.create(<Crossword {...defaultProps} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('creates new gridData when the data changes', () => {
  const clueMatch = /apple/;
  const { queryByText, rerender } = render(<Crossword {...defaultProps} />);
  expect(queryByText(clueMatch)).toBeNull();

  rerender(<Crossword {...defaultProps} data={simpleData} />);
  expect(queryByText(clueMatch)).toBeTruthy();
});

describe('imperative commands (forwarded to CrosswordProvider)', () => {
  it('sets focus when requested', () => {
    const ref = React.createRef<CrosswordImperative>();
    const { getByLabelText, container } = render(
      <Crossword {...defaultProps} data={simpleData} ref={ref} />
    );

    const doc = container.ownerDocument;

    // no focus yet?
    const input = getByLabelText('crossword-input');
    expect(doc.activeElement).not.toBe(input);

    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.focus();
    });

    expect(doc.activeElement).toBe(input);
  });

  it('resets data when requested', () => {
    const ref = React.createRef<CrosswordImperative>();
    const { getByLabelText, queryByText } = render(
      <Crossword {...defaultProps} data={simpleData} ref={ref} />
    );

    userEvent.click(getByLabelText('clue-1-across'));

    const input = getByLabelText('crossword-input');

    userEvent.paste(input, 'い');
    let textEl = queryByText('い');
    expect(textEl).toBeTruthy();

    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.reset();
    });

    textEl = queryByText('い');
    expect(textEl).toBeFalsy();
  });

  it('fills answers when requested', () => {
    const ref = React.createRef<CrosswordImperative>();
    const { queryByText } = render(
      <Crossword {...defaultProps} data={simpleData} ref={ref} />
    );

    let textEl = queryByText('ご');
    expect(textEl).toBeFalsy();

    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.fillAllAnswers();
    });

    textEl = queryByText('ご');
    expect(textEl).toBeTruthy();
  });

  it('calls onLoadedCorrect after filling answers', () => {
    const onLoadedCorrect = jest.fn();
    const ref = React.createRef<CrosswordImperative>();
    render(
      <Crossword
        {...defaultProps}
        data={simpleData}
        onLoadedCorrect={onLoadedCorrect}
        ref={ref}
      />
    );

    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.fillAllAnswers();
    });

    expect(onLoadedCorrect).toBeCalledWith([
      ['across', '1', 'りんご'],
      ['down', '2', 'ごりら'],
    ]);
  });

  it('calls onCrosswordCorrect after filling answers', () => {
    const onCrosswordCorrect = jest.fn();
    const ref = React.createRef<CrosswordImperative>();
    render(
      <Crossword
        {...defaultProps}
        data={simpleData}
        onCrosswordCorrect={onCrosswordCorrect}
        ref={ref}
      />
    );

    onCrosswordCorrect.mockClear();
    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.fillAllAnswers();
    });

    expect(onCrosswordCorrect).toBeCalledTimes(1);
    expect(onCrosswordCorrect).toBeCalledWith(true);
  });

  it('returns whether the crossword is correct', () => {
    const ref = React.createRef<CrosswordImperative>();
    render(<Crossword {...defaultProps} data={simpleData} ref={ref} />);
    expect(ref.current).toBeTruthy();
    let isCorrect = true;
    act(() => {
      isCorrect = ref.current?.isCrosswordCorrect();
    });

    expect(isCorrect).toBeFalsy();
  });

  it('setGuess() can set a guess', () => {
    const ref = React.createRef<CrosswordImperative>();
    const { queryByText } = render(
      <Crossword {...defaultProps} data={simpleData} ref={ref} />
    );

    let textEl = queryByText('り');
    expect(textEl).toBeFalsy();

    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.setGuess(0, 0, 'り');
    });

    textEl = queryByText('り');
    expect(textEl).toBeTruthy();
  });
});
