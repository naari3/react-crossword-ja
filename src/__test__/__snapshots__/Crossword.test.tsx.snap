// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Array [
  <div
    className="sc-bdvvtL fECXGP crossword grid"
  >
    <div
      style={
        Object {
          "margin": 0,
          "padding": 0,
          "position": "relative",
        }
      }
    >
      <svg
        viewBox="0 0 100 100"
      >
        <rect
          fill="rgb(0,0,0)"
          height={100}
          width={100}
          x={0}
          y={0}
        />
      </svg>
      <input
        aria-label="crossword-input"
        autoComplete="off"
        autoCorrect="off"
        onChange={[Function]}
        onClick={[Function]}
        onCompositionEnd={[Function]}
        onKeyDown={[Function]}
        spellCheck="false"
        style={
          Object {
            "backgroundColor": "transparent",
            "border": 0,
            "caretColor": "transparent",
            "color": "rgba(0, 0, 0, 0)",
            "cursor": "default",
            "fontSize": "418.94999999999993px",
            "height": "calc(100% - 4px)",
            "left": "calc(0% + 2px)",
            "margin": 0,
            "padding": 0,
            "position": "absolute",
            "textAlign": "center",
            "textAnchor": "middle",
            "top": "calc(0% + 2px)",
            "width": "calc(100% - 4px)",
          }
        }
        type="text"
        value=""
      />
    </div>
  </div>,
  <div
    className="sc-dkPtRN fFCnQE clues"
  >
    <div
      className="direction"
    >
      <h3
        className="header"
      >
        ACROSS
      </h3>
    </div>
    <div
      className="direction"
    >
      <h3
        className="header"
      >
        DOWN
      </h3>
    </div>
  </div>,
]
`;
